packages/@lwc/engine-core/src/framework/rendering.ts

```diff
@@ -976,9 +969,28 @@ function updateStaticChildren(c1: VNodes, c2: VNodes, parent: ParentNode, render
         if (n2 !== n1) {
             if (isVNode(n1)) {
                 if (isVNode(n2)) {
-                    // both vnodes are equivalent, and we just need to patch them
-                    // note: dynamic components can be patched in this call
-                    patch(n1, n2, parent, renderer);
+                    if (isSameVnode(n1, n2)) {
+                        // both vnodes are equivalent, and we just need to patch them
+                        patch(n1, n2, parent, renderer);
+                    } else {
+                        // In the case of dynamic components, the element will always occupy the same position in the DOM
+                        // as specified by lwc:component. In such cases, we need to unmount the existing element and
+                        // mount the new one using the new constructor.
+                        if (
+                            n1.type === VNodeType.CustomElement &&
+                            n2.type === VNodeType.CustomElement
+                        ) {
+                            if (n1.ctor !== n2.ctor) {
+                                // If the constructors differ, unmount the current component and mount a new one using
+                                // the new constructor. This scenario only occurs for dynamic components.
+                                const dynElmRenderer = n2.data.renderer ?? renderer;
+                                const dynElmAnchor = dynElmRenderer.nextSibling(n1.elm);
+
+                                unmount(n1, parent, dynElmRenderer, true);
+                                mountCustomElement(n2, parent, dynElmAnchor, dynElmRenderer);
+                            }
+                        }
+                    }
```

Comment:
This logic comes straight from `patchCustomElements` and is the [logic](https://github.com/salesforce/lwc/blob/master/packages/%40lwc/engine-core/src/framework/rendering.ts#L380-L386) that runs in the current diffing algo logic for dynamic components.

This logic was needed for `lwc:dynamic` because in the dynamic diffing algo the checks for [`isSameVnode`](https://github.com/salesforce/lwc/blob/master/packages/%40lwc/engine-core/src/framework/rendering.ts#L884) would always trigger true.  Thus if a new constructor was supplied, it would run `patchCustomElement` and need to unmount and mount the new component using the same tag name.

In the new implementation, `isSameVnode` will never be true unless the dynamic component changes (the value of `sel` will be different).

Additionally, with the changes made in #2966 the dynamic components can now travel through the static path.  This is allowed because we now consider the dynamic component before and after the constructor is changed to be [equivalent](https://github.com/salesforce/lwc/blob/master/packages/%40lwc/engine-core/src/framework/rendering.ts#L970-L971).

Having said that, I've moved this logic out of `patchCustomElement` into `updateStaticChildren` because this path is still triggered when the constructor changes for a dynamic component.  With the new implementation, when the constructors change, the dynamic components vnodes should still be considered equivalent (as they are in `lwc:dynamic`).

The logic is not needed in `patchCustomElement` anymore because the part of the dynamic diffing algo that relies on [isSameVnode](https://github.com/salesforce/lwc/blob/master/packages/%40lwc/engine-core/src/framework/rendering.ts#L884) will not trigger true when the dynamic component's constructor changes in the new implmentation.

An alternative approach could be to emit a fragment for each dynamic component but I chose to avoid this because each fragment is wrapped in empty text nodes.  I felt that the empty text nodes were unnecessary in this case since there will only ever be a single node.

Reply:
@pmdartus  would love to get your thoughts on the :point_up: if you get a chance 😃

Reply:
> The logic is not needed in `patchCustomElement` anymore because the part of the dynamic diffing algo that relies on [isSameVnode](https://github.com/salesforce/lwc/blob/master/packages/%40lwc/engine-core/src/framework/rendering.ts#L884) will not trigger true when the dynamic component's constructor changes in the new implmentation.

I don't fully understand what you mean by the "the part of the dynamic diffing algo that relies on isSameVnode".

Overall, I don't think the dynamic component diffing logic belongs to the `updateStaticChildren`. How custom elements are patched should be an internal detail of the `patchCustomElement`.

Reply:
> I don't fully understand what you mean by the "the part of the dynamic diffing algo that relies on isSameVnode".

@pmdartus  sorry I was mistaken about this part, I've updated the comment. 

> Maybe the question is: Why is this code even needed?

That's a great observation, yes I think the issue is the diffing algo hasn't had to deal with diffing based on different tags (at least in `patch`).

> I.e. why check that they are both custom elements? If one is a div and the other is an input, we still want to completely unmount one and mount the other one. We do need to check if the second is a custom element, though, so we can call mountCustomElement instead of mountElement.

We're doing this in `patchCustomElement` right now but I think you bring up an interesting point.  Should we expand `patch` to also unmount / mount nodes when the 2 node types `n1` and `n2` do not have the same tag name?  Or alternatively, create a new routine specifically for replacing nodes when the tag name does not match?

For now, since the only scenario where the tag name can differ is for custom elements, I've reverted the change I made and moved the logic back into `patchCustomElement`.  

I've also updated the dev error here since patching a dynamic components no longer have the same tag names.

Reply:
@nolanlawson based on your feedback I updated the unmounting / mounting [logic](https://github.com/salesforce/lwc/pull/3338/commits/1398300d6a927596f53ab972d700d8960554f320) in `updateStaticChildren` to be agnostic of the type that is a mismatch.

With this implementation, `patch` now serves the sole purpose of only updating the same element.

Let me know what you think, I'm not sure if this is a breaking change or not.  My thoughts are that it shouldn't be.

Reply:
If the tests still pass, then it's probably not a breaking change. Nucleus also tends to catch diffing algo changes that cause downstream breakages.

Reply:
Based on the review this morning, the team decided to isolate the changes in unmounting / mounting of the dynamic component to `patchCustomElement`.

In a future PR, we will introduce the unmounting / mounting in `updateStaticChildren` to help isolate any potential future regression.